// @impactdk/barcode-scanner/esm2015/src/decoder/decoder.js
import rawr from "rawr";
import transport from "rawr/transports/worker";
const workerFileName = "wasm-worker.js";
function sanitizeDirString(s) {
    return s.replace(/\/$/, "");
}
export class WasmDecoder {
    constructor(installationDir = "") {
        const workerFilePath = `${installationDir}/${workerFileName}`;
        this.worker = new Worker(workerFilePath);
        this.wasmDecoder = rawr({ transport: transport(this.worker) });
    }
    static getInstance(installationDir = "") {
        if (!WasmDecoder.instance) {
            const sanitized = sanitizeDirString(installationDir);
            WasmDecoder.instance = new WasmDecoder(sanitized);
        }
        return WasmDecoder.instance;
    }
    static removeInstance() {
        WasmDecoder.instance.dispose();
        WasmDecoder.instance = null;
    }
    decode(context) {
        const canvas = context.canvas;
        const width = canvas.width;
        const height = canvas.height;
        const imageData = context.getImageData(0, 0, width, height);
        return this.wasmDecoder.methods.detectUrl(width, height, imageData);
    }
    dispose() {
        this.wasmDecoder = null;
        this.worker.terminate();
    }
}
