// @impactdk/react-barcode-scanner/fesm5/impactdk-react-barcode-scanner.js
import { __assign } from 'tslib';
import React, { forwardRef, useRef, useEffect, useImperativeHandle } from 'react';
import { Scanner } from '@impactdk/barcode-scanner';

var videoId = "scanner-video";
/**
 * @example
 * import { WasmDecoder } from "@impactdk/barcode-scanner";
 * import { ReactBarcodeScanner, IReactBarcodeScannerRef } from "@impactdk/react-barcode-scanner";
 *
 * const scannerRef = useRef<IReactBarcodeScannerRef>(); // Access to public API, i.e. scannerRef?.start() or scannerRef?.stop().
 *
 * const decoder = WasmDecoder.getInstance("/public-path/asssets/barcode-scanner"); // Path to installed assets.
 *
 * return <ReactBarcodeScanner decoder={decoder} onFindBarcode={setBarcode} ref={scannerRef} />;
 */
var ReactBarcodeScanner = forwardRef(function (props, ref) {
    var videoProps = props.videoProps, onFindBarcode = props.onFindBarcode, decoder = props.decoder, _a = props.initOnMount, initOnMount = _a === void 0 ? true : _a;
    var scannerRef = useRef(null);
    useEffect(function () {
        if (initOnMount) {
            startScanner();
        }
        return handleUnmount;
    }, []);
    useImperativeHandle(ref, function () { return ({
        start: startScanner,
        stop: stopScanner
    }); });
    function startScanner() {
        if (scannerRef.current) {
            return;
        }
        var video = document.getElementById(videoId);
        scannerRef.current = new Scanner(video, decoder, (function () {
            var previousValue;
            return function (_a) {
                var rawValue = _a.rawValue;
                // Compare with previous value, to ensure the correct one has been found.
                if (!previousValue || rawValue !== previousValue) {
                    previousValue = rawValue;
                    return;
                }
                onFindBarcode(rawValue);
            };
        })());
        scannerRef.current.start();
    }
    function stopScanner() {
        if (scannerRef.current != null) {
            scannerRef.current.stop();
            scannerRef.current = null;
        }
    }
    function handleUnmount() {
        stopScanner();
    }
    return (React.createElement("video", __assign({}, videoProps, { id: videoId })));
});

export { ReactBarcodeScanner };
