// @impactdk/react-barcode-scanner/fesm2015/impactdk-react-barcode-scanner.js
import React, { forwardRef, useRef, useEffect, useImperativeHandle } from 'react';
import { Scanner } from '@impactdk/barcode-scanner';

const videoId = "scanner-video";
/**
 * @example
 * import { WasmDecoder } from "@impactdk/barcode-scanner";
 * import { ReactBarcodeScanner, IReactBarcodeScannerRef } from "@impactdk/react-barcode-scanner";
 *
 * const scannerRef = useRef<IReactBarcodeScannerRef>(); // Access to public API, i.e. scannerRef?.start() or scannerRef?.stop().
 *
 * const decoder = WasmDecoder.getInstance("/public-path/asssets/barcode-scanner"); // Path to installed assets.
 *
 * return <ReactBarcodeScanner decoder={decoder} onFindBarcode={setBarcode} ref={scannerRef} />;
 */
const ReactBarcodeScanner = forwardRef((props, ref) => {
    const { videoProps, onFindBarcode, decoder, initOnMount = true } = props;
    const scannerRef = useRef(null);
    useEffect(() => {
        if (initOnMount) {
            startScanner();
        }
        return handleUnmount;
    }, []);
    useImperativeHandle(ref, () => ({
        start: startScanner,
        stop: stopScanner
    }));
    function startScanner() {
        if (scannerRef.current) {
            return;
        }
        const video = document.getElementById(videoId);
        scannerRef.current = new Scanner(video, decoder, (() => {
            let previousValue;
            return ({ rawValue }) => {
                // Compare with previous value, to ensure the correct one has been found.
                if (!previousValue || rawValue !== previousValue) {
                    previousValue = rawValue;
                    return;
                }
                onFindBarcode(rawValue);
            };
        })());
        scannerRef.current.start();
    }
    function stopScanner() {
        if (scannerRef.current != null) {
            scannerRef.current.stop();
            scannerRef.current = null;
        }
    }
    function handleUnmount() {
        stopScanner();
    }
    return (React.createElement("video", Object.assign({}, videoProps, { id: videoId })));
});

export { ReactBarcodeScanner };
