// @impactdk/barcode-scanner/esm5/src/decoder/decoder.js
import rawr from "rawr";
import transport from "rawr/transports/worker";
var workerFileName = "wasm-worker.js";
function sanitizeDirString(s) {
    return s.replace(/\/$/, "");
}
var WasmDecoder = /** @class */ (function () {
    function WasmDecoder(installationDir) {
        if (installationDir === void 0) { installationDir = ""; }
        var workerFilePath = installationDir + "/" + workerFileName;
        this.worker = new Worker(workerFilePath);
        this.wasmDecoder = rawr({ transport: transport(this.worker) });
    }
    WasmDecoder.getInstance = function (installationDir) {
        if (installationDir === void 0) { installationDir = ""; }
        if (!WasmDecoder.instance) {
            var sanitized = sanitizeDirString(installationDir);
            WasmDecoder.instance = new WasmDecoder(sanitized);
        }
        return WasmDecoder.instance;
    };
    WasmDecoder.removeInstance = function () {
        WasmDecoder.instance.dispose();
        WasmDecoder.instance = null;
    };
    WasmDecoder.prototype.decode = function (context) {
        var canvas = context.canvas;
        var width = canvas.width;
        var height = canvas.height;
        var imageData = context.getImageData(0, 0, width, height);
        return this.wasmDecoder.methods.detectUrl(width, height, imageData);
    };
    WasmDecoder.prototype.dispose = function () {
        this.wasmDecoder = null;
        this.worker.terminate();
    };
    return WasmDecoder;
}());
export { WasmDecoder };
